authui/src/authflowv2/loading.ts

```diff
@@ -0,0 +1,63 @@
+import { Controller } from "@hotwired/stimulus";
+import {
+  disableAllButtons,
+  hideProgressBar,
+  showProgressBar,
+} from "../loading";
+
+interface LoadingEvent extends Event {
+  detail?: {
+    button?: HTMLButtonElement;
+    error?: unknown;
+  };
+}
+
+export class LoadingController extends Controller {
+  revertDisabledButtons: { (): void } | null = null;
+  button: HTMLButtonElement | null = null;
+
+  // Revert disabled buttons before Turbo caches the page
+  // To avoid flickering in the UI
+  beforeCache = () => {
+    if (this.revertDisabledButtons) {
+      this.revertDisabledButtons();
+    }
+  };
+
+  onLoading(e: LoadingEvent) {
+    this.revertDisabledButtons = disableAllButtons();
+    showProgressBar();
+    if (e.detail?.button?.getAttribute("data-loading-state") != null) {
+      e.detail.button.classList.add("primary-btn--loading");
```

Comment:
Do not write `primary-btn--loading` here, because this coupled this controller with primary button.
I expect it could be an attribute `loading` or `data-loading` added to the button, which doesn't care about the button type.

The component (primary button) should handle this attribute change in his own code (We can do it in the stylesheet)